src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs

```diff
@@ -23,65 +25,239 @@ namespace System.Text.RegularExpressions.Generator
+            "#pragma warning disable CS0164 // Unreferenced label",
+            "#pragma warning disable CS0219 // Variable assigned but never used",
+        };
 
-            // Contains one entry per regex method, either the generated code for that regex method,
-            // a diagnostic to fail with, or null if no action should be taken for that regex.
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            // Produces one entry per generated regex.  This may be:
+            // - Diagnostic in the case of a failure that should end the compilation
+            // - (RegexType regexType, string runnerFactoryImplementation, Dictionary<string, string[]> requiredHelpers) in the case of valid regex
+            // - (RegexType regexType, string reason, Diagnostic diagnostic) in the case of a limited-support regex
             IncrementalValueProvider<ImmutableArray<object?>> codeOrDiagnostics =
                 context.SyntaxProvider
 
-                // Find all MethodDeclarationSyntax nodes attributed with RegexGenerator and gather the required information
+                // Find all MethodDeclarationSyntax nodes attributed with RegexGenerator and gather the required information.
                 .CreateSyntaxProvider(IsSyntaxTargetForGeneration, GetSemanticTargetForGeneration)
                 .Where(static m => m is not null)
 
-                // Pair each with whether unsafe code is allowed
-                .Combine(allowUnsafeProvider)
-
-                // Get the resulting code string or error Diagnostic for
-                // each MethodDeclarationSyntax/allow-unsafe-blocks pair
+                // Generate the RunnerFactory for each regex, if possible.  This is where the bulk of the implementation occurs.
                 .Select((state, _) =>
                 {
-                    Debug.Assert(state.Left is not null);
-                    return state.Left is RegexType regexType ? EmitRegexType(regexType, state.Right) : state.Left;
+                    if (state is not RegexType regexType)
+                    {
+                        Debug.Assert(state is Diagnostic);
+                        return state;
+                    }
+
+                    // If we're unable to generate a full implementation for this regex, report a diagnostic.
+                    // We'll still output a limited implementation that just caches a new Regex(...).
+                    if (!regexType.Method.Tree.Root.SupportsCompilation(out string? reason))
+                    {
+                        return (regexType, reason, Diagnostic.Create(DiagnosticDescriptors.LimitedSourceGeneration, regexType.Method.MethodSyntax.GetLocation()));
+                    }
+
+                    // Generate the core logic for the regex.
+                    Dictionary<string, string[]> requiredHelpers = new();
+                    var sw = new StringWriter();
```

Comment:
I'm sure I asked this before, but shouldn't you dispose the writer(s)?, I know the dispose is mostly cosmetics (setting IsOpen to false)

Reply:
`StringWriter` does not hold any native resources such as files. It is simply a wrapper over `StringBuilder`, and has `Dispose` because `TextWriter` has.

Reply:
Yes, I'm aware that the dispose is not doing much in this case, as noted above and I'm aware that the IndentedWriter is legacy codedom code, so it likely will never change. Still what is the downside of adding two using keywords here, for the off chance that somebody will add eg StringBuilder Pooling to it, as with more source generators, more people will hopefully use the IndentedWriter than writing their own weird indention logic.

Reply:
`new StringWriter()` will never add such behavior for compatibility reasons. This would be either a `PooledStringWriter` type or a new constructor overload.